tests/hawc/apps/epi/test_epi_api.py

```diff
@@ -71,3 +76,1652 @@ def test_result_heatmap(self, rewrite_data_files: bool, db_keys):
+                "data": {
+                    "study": db_keys.study_working,
+                    "eligible_n": "not_numeric",
+                    "invited_n": {},
+                    "participant_n": False,
+                },
+            },
+        )
+
+        generic_test_scenarios(client, url, scenarios)
+
+    def test_valid_requests(self, db_keys):
+        url = reverse("epi:api:study-population-list")
+        client = APIClient()
+        assert client.login(username="team@hawcproject.org", password="pw") is True
+
+        # Make some test criteria
+        assessment = Assessment.objects.get(id=db_keys.assessment_working)
+        criteria1 = models.Criteria(assessment=assessment, description="dummy 1")
+        criteria1.save()
+
+        criteria2 = models.Criteria(assessment=assessment, description="dummy 2")
+        criteria2.save()
+
+        base_data = {
+            "study": db_keys.study_working,
+            "name": "study pop name",
+            "design": "CO",
+            "age_profile": "test profile",
+            "source": "test source",
+            "countries": ["JP"],
+            "region": "test region",
+            "state": "test state",
+            "eligible_n": 3,
+            "invited_n": 2,
+            "participant_n": 1,
+            "comments": "test comments",
+        }
+
+        just_created_study_id = None
+
+        def study_pop_lookup_test(resp):
+            nonlocal just_created_study_id
+
+            study_pop_id = resp.json()["id"]
+            study_pop = models.StudyPopulation.objects.get(id=study_pop_id)
+            assert study_pop.design == "CO"
+            assert study_pop.name == base_data["name"]
+            # Etc...
+
+            if just_created_study_id is None:
+                just_created_study_id = study_pop_id
+
+        def study_pop_lookup_test_with_criteria(resp):
+            study_pop_id = resp.json()["id"]
+            study_pop = models.StudyPopulation.objects.get(id=study_pop_id)
+
+            found_c1 = False
+            found_c2 = False
+            found_c3 = False
+            for c in study_pop.criteria.all():
+                if c.id == criteria1.id:
+                    found_c1 = True
+                elif c.id == criteria2.id:
+                    found_c2 = True
+                elif c.description == "on the fly":
+                    found_c3 = True
+
+            assert found_c1 is True and found_c2 is True and found_c3 is True
+
+        def altered_study_pop_test(resp):
+            nonlocal just_created_study_id
+
+            study_pop_id = resp.json()["id"]
+            study_pop = models.StudyPopulation.objects.get(id=study_pop_id)
+            assert study_pop.name == "updated"
+            assert study_pop_id == just_created_study_id
+
+        def deleted_study_pop_test(resp):
+            nonlocal just_created_study_id
+
+            assert resp.data is None
+            try:
```

Comment:
rewrite this type of thing in your test to 

```python
with pytest.raises(ObjectDoesNotExistError):
    models.StudyPopulation.objects.get(id=just_created_study_id)
```

https://docs.pytest.org/en/stable/reference.html#pytest.raises

Reply:
neat, didn't know about that one.